lua/plenary/async_lib/utils.lua

```diff
@@ -0,0 +1,233 @@
+local a = require('plenary.async_lib.async')
+local await = a.await
+local async = a.async
+local co = coroutine
+local VecDeque = require('plenary.async_lib.helpers').VecDeque
+local uv = vim.loop
+
+local M = {}
+
+M.sleep = a.wrap(function(ms, callback)
+  local timer = uv.new_timer()
+  uv.timer_start(timer, ms, 0, function()
+    uv.timer_stop(timer)
+    uv.close(timer)
+    callback()
+  end)
+end, 2)
+
+--- Takes a future and a millisecond as the timeout.
+--- If the time is reached and the future hasn't completed yet, it will short circuit the future
+--- NOTE: the future will still be running in libuv, we are just not waiting for it to complete
+--- thats why you should call this on a leaf future only to avoid unexpected results
+M.timeout = a.wrap(function(future, ms, callback)
+  -- make sure that the callback isn't called twice, or else the coroutine can be dead
+  local done = false
+
+  local timeout_callback = function(...)
+    if not done then
+      done = true
+      callback(false, ...) -- false because it has run normally
+    end
+  end
+
+  vim.defer_fn(function()
+    if not done then
+      done = true
+      callback(true) -- true because it has timed out
+    end
+  end, ms)
+
+  a.run(future, timeout_callback)
+end, 3)
+
+M.timer = function(ms)
+  return async(function()
+    await(M.sleep(ms))
+  end)
+end
+
+M.id = async(function(...)
+  return ...
+end)
+
+M.thread_loop = function(thread, callback)
+  local idle = uv.new_idle()
+  idle:start(function()
+    local success = co.resume(thread)
+    assert(success, "Coroutine failed")
+
+    if co.status(thread) == "dead" then
+      idle:stop()
+      callback()
+    end
+  end)
+end
+
+M.thread_loop_async = a.wrap(M.thread_loop, 2)
+
+M.yield_now = async(function()
+  await(M.id())
+end)
+
+-- will force the future to be runned when called
+M.runned = function(future)
+  return function(step)
+    a.run(future, step)
+  end
+end
+
+local Condvar = {}
+Condvar.__index = Condvar
+
+function Condvar.new()
+  return setmetatable({handles = {}}, Condvar)
+end
+
+--- async function
+--- blocks the thread until a notification is received
+Condvar.wait = a.wrap(function(self, callback)
+  -- not calling the callback will block the coroutine
+  table.insert(self.handles, callback)
+end, 2)
+
+--- not an async function
+function Condvar:notify_all()
```

Comment:
perhaps we should prepend `sync_` to method names that are not async. You want the API to be as discoverable as possible via name, since that will make reading the corresponding code much easier.

Reply:
I agree that we should change the name for functions that are synchronous. However, I don't think that we should do that in this case because it is just a regular functions. The synchronous utility functions are just for control flow and don't actually do any io, so I think it would be confusing to prepend `sync_`. I would do this for functions that actually do things synchronous and use io, like our `Path` library or the `curl` library. For example `Path:sync_mkdir` or `Path:sync_rename`.